### ISSUE TITLE
Snapshot restore throws UPnP error 402

### ISSUE THREAD

**Issue by Dan1001** [SoCo/SoCo#761]
Hi,

multi_zone_snap.py consistently fails if any of the Sonos devices is playing music - upon the restore, it gives a 402 error as follows:

`Traceback (most recent call last):
  File "./test2.py", line 84, in <module>
    play_alert(
  File "./test2.py", line 74, in play_alert
    zone.snap.restore(fade=fade_back)
  File "/usr/local/lib/python3.8/dist-packages/soco/snapshot.py", line 214, in restore
    self.device.play_uri(
  File "/usr/local/lib/python3.8/dist-packages/soco/core.py", line 121, in inner_function
    return function(self, *args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/soco/core.py", line 642, in play_uri
    self.avTransport.SetAVTransportURI(
  File "/usr/local/lib/python3.8/dist-packages/soco/services.py", line 217, in _dispatcher
    return self.send_command(action, *args, **kwargs)
  File "/usr/local/lib/python3.8/dist-packages/soco/services.py", line 507, in send_command
    self.handle_upnp_error(response.text)
  File "/usr/local/lib/python3.8/dist-packages/soco/services.py", line 560, in handle_upnp_error
    raise SoCoUPnPException(
soco.exceptions.SoCoUPnPException: UPnP Error 402 received: Invalid Args from 192.168.1.30
`

(with 192.168.1.30 being the Sonos devices that was playing at the time)

thanks!

Dan

**Comment by DPH** [comment]
Hi @Dan1001 
I have tested out the example program in the SoCo release with Python 3.8 my system is S1 Sonos.
I can't make it fail whatever is playing or not!

Looking at the error you get it implies you are playing a stream / radio station - can you let me know what you are playing please. Also are you on S1 or S2 sonos? Assume you are using SoCo 0.20 ?

Maybe you could try the example in the SoCo release https://github.com/SoCo/SoCo/blob/master/examples/snapshot/multi_zone_snap.py
and see if that still fails for you. 

Let us know how you get on and we can investigate further,
Cheers David

**Comment by Dan1001** [comment]
Many thanks. I'm guessing the difference is that my system is S2 Sonos.

My SoCo release is 0.20. Python 3.8.5, and I was playing a stream on Spotify, then testing using an unedited version of multi_zone_snap.py. 

Thanks for the pointers to the other issues around 702. I tried the suggestion of amending "pause" to "stop" in line 178 of snapshot.py - that doesn't help.

Dan

**Comment by DPH** [comment]
Interesting. How many zones are you snapshotting, how many are grouped and how many are playing music?
Which zone is it failing on - the device - maybe its something special on that device.
Assume your alert is as in the example.
Can anyone else try the example on an S2 set up please?
Cheers David

**Comment by Dan1001** [comment]
I've tested, and get the error whether I run the alert (per the example) on all zones, or on individual zones one at a time (ungrouped). All the devices are One SLs, and all respond in the same way.

Some more testing: the error appears consistently when streaming from Spotify on either my iphone or macbook. However there is no error if streaming from youtube on iphone/macbook. And no error if Sonos itself is streaming from Spotify (i.e. using the Sonos app not the Spotify app).

Dan

**Comment by pwt** [comment]
> Some more testing: the error appears consistently when streaming from Spotify on either my iphone or macbook. However there is no error if streaming from youtube on iphone/macbook. And no error if Sonos itself is streaming from Spotify (i.e. using the Sonos app not the Spotify app).

It sounds like you're streaming to the Sonos speakers using AirPlay?

**Comment by pwt** [comment]
> Can anyone else try the example on an S2 set up please?
> Cheers David

Working fine for me on S2, playing Spotify directly from Sonos.

**Comment by Dan1001** [comment]
It's Spotify Connect

**Comment by Dan1001** [comment]
@pwt, could you please try if you play Spotify from the Spotify app?

**Comment by pwt** [comment]
> @pwt, could you please try if you play Spotify from the Spotify app?

I get the same error as you, if using Spotify Connect. However, I think that's what I'd expect when Sonos is being controlled by something external to the system. I'd expect the same limitation with AirPlay playback.

**Comment by Dan1001** [comment]
oh - perhaps this is inevitable, then!

**Comment by DPH** [comment]
Playing this way erros on S1 also.  Like queues started from Alexa or Google Home (cloud queues) they can't currently be restarted from SoCo. You can pause  them and play them, but once you stop or play an alert sound you loose connection with the original stream. 

There are some options:
- we can trap this stream and not raise an error -  but it will just not restart - as with other cloud queues. if you want to try it insert the following at line 121 in (indented same level as previous if)
```
        if self.media_uri.split(":")[0] == "x-sonos-vli":
            # playing from an external service (eg via Spotify connect)
            self.is_playing_cloud_queue = True
```

- in you code you could pause this stream leave one player with it - alert on the others players then rejoin the other players to the preserved player and play. We could make this an addition to snapshot to have a "preserve cloud queue player" which would work for this and other cloud queues.

Welcome your thoughts or other ideas
Cheers David

**Comment by peterwallhead** [comment]
I'm hitting this same issue now with Spotify Connect. I've tried the the line 121 addition suggestion above, but it still disconnects Spotify and fails to reconnect on restore. Any more progress since this time last year please?

**Comment by pwt** [comment]
> @DPH, I'm hitting this same issue now with Spotify Connect. I've tried the the line 121 addition suggestion above, but it still disconnects Spotify and fails to reconnect on restore. Any more progress since this time last year please?

I'm not sure there is an acceptably clean solution to this use case.

**Comment by DPH** [comment]
Hi @peterwallhead, I think the only way you will resume Spotify Connect is if you use a lib that controls Spotify and that casts to Sonos. As said before if a queue is created outside of Sonos (Alexa, Spotify etc.) then Sonos has no knowledge of it and no ability to restart it. Please post any options you find as I think a lot of us are interested in making this work. Good luck, Cheers David

**Comment by peterwallhead** [comment]
Ah, thanks @DPH! That's definitely helped my understanding of what's going on. I was focussing on controlling the Sonos device, but I now understand that I should be controlling Spotify instead in this instance.

Basically I'll just need to do the following, right?

- Snapshot the current Spotify Connect state (and then stop the music), all via Spotify
- Take control of the device using SoCo (and play the required alert/message)
- Restore the Spotify state by reconnecting/recasting to the device via Spotify

Will post here if/when I find a workable solution.

**Comment by DPH** [comment]
Sounds right to me. Will be great if you can get that combination working.
Happy to test things or help if you find some options.

Cheers David

**Comment by emcniece** [comment]
This thread got locked, so I'll reference it from this end: https://github.com/home-assistant/core/issues/45050

Tl;dr: Home Assistant is also affected by this issue when restoring a snapshot taken while a Spotify stream was playing.
